sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala

```diff
@@ -50,8 +50,25 @@ object PropagateEmptyRelation extends Rule[LogicalPlan] with PredicateHelper wit
   override def conf: SQLConf = SQLConf.get
 
   def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
-    case p: Union if p.children.forall(isEmptyLocalRelation) =>
-      empty(p)
+    case p @ Union(children) if children.exists(isEmptyLocalRelation) =>
+      val newChildren = children.filterNot(isEmptyLocalRelation)
+      if (newChildren.isEmpty) {
+        empty(p)
+      } else {
+        val newFirstChild = if (newChildren.head eq children.head) {
+          newChildren.head
+        } else {
+          val pl = children.head.output.zip(newChildren.head.output).map {
+            case (oa, na) => Alias(na, oa.name)(oa.exprId)
+          }
+          Project(pl, newChildren.head)
+        }
```

Comment:
Looks like this may change output nullability of original Union?

Reply:
but it's corrected?

Reply:
This project looks correct. Just the new Union's output nullability might change.

Reply:
Maybe a stupid question, but is that an issue? It can change in one direction, from nullable to non-nullable.

Reply:
It should not be an issue. But the nullability updates may not be effective unless we run the rule `UpdateAttributeNullability`

Reply:
Ah ok, I see. So shall we run `UpdateAttributeNullability` after `PropagateEmptyRelation` in the `Optimizer`?

Reply:
currently `UpdateAttributeNullability` is put in a `Once` batch. We need to evaluate if it's OK (idempotent and cost-efficient) to run it in a fix-point batch.

Reply:
It looks like it is idempotent and doesn't do costly things just a simple bottom-up traversal on nodes and one expression traversal per node.

Reply:
I've added `UpdateAttributeNullability` after `PropagateEmptyRelation` in the latest commit.